README.md
<div align="center">
    
<img src="https://raw.githubusercontent.com/ietf-tools/common/main/assets/logos/pypi-publish.svg" alt="PYPI PUBLISH" height="125" />

[![Release](https://img.shields.io/github/release/ietf-tools/pypi-publish.svg?style=flat&maxAge=600)](https://github.com/ietf-tools/pypi-publish/releases)
[![License](https://img.shields.io/github/license/ietf-tools/pypi-publish)](https://github.com/ietf-tools/pypi-publish/blob/main/LICENSE)
[![npm](https://img.shields.io/npm/v/@ietf-tools/pypi-publish)](https://www.npmjs.com/package/@ietf-tools/pypi-publish)
[![node-current](https://img.shields.io/node/v/@ietf-tools/pypi-publish)](https://github.com/ietf-tools/pypi-publish)
    
##### Tool for publishing a Python package to PyPI from a GitHub Release
    
</div>

- [Changelog](https://github.com/ietf-tools/pypi-publish/blob/main/CHANGELOG.md)
- [Contributing](https://github.com/ietf-tools/.github/blob/main/CONTRIBUTING.md)
- [Requirements](#requirements)
- [Usage](#usage)

---

This tool is a CLI which provides the following automation:

- Fetch the list of available repositories and releases
- Download the latest build of a Python package
- Install Twine *(if not already installed)*
- Sign and publish the package to PyPI (or TestPyPI)

## Requirements

- [Node.js](https://nodejs.org/) **16.x or later**
- [Python](https://www.python.org/) **3.x**

> This tool assumes that you have the signing key used to sign Python packages already configured on your system. It will be used when publishing the package to PyPI.

## Usage

Install the `@ietf-tools/pypi-publish` NPM package globally using:

```sh
npm install -g @ietf-tools/pypi-publish
```

Then run *(from any location)*:

```sh
pypi-publish
```

Enter the necessary info as prompted.

### CLI Arguments *(optional)*

These arguments can also be passed to the CLI to automate values and bypass the questions. All arguments are optional.

| Short         | Long                  | Description                                 |
|---------------|-----------------------|---------------------------------------------|
| `-t TARGET`   | `--target=TARGET`     | Target PyPI repository [`pypi`, `testpypi`] |
| `-a TOKEN`    | `--token=TOKEN`       | PyPI API Token                              |
| `-i IDENTITY` | `--identity=IDENTITY` | GPG identity to use for package signing     |
| `-p PROJECT`  | `--project=PROJECT`   | GitHub project (repository) to publish from |
| `-r RELEASE`  | `--release=RELEASE`   | GitHub release to publish                   |
|               | `--python-path=PATH`  | Path to Python executable                   |
| `-h`          | `--help`              | Display usage + help message and exit       |
| `-v`          | `--version`           | Display CLI version and exit                |

## License

BSD-3-Clause


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [v1.1.0] - 2022-08-29
### :sparkles: New Features
- [`a020065`](https://github.com/ietf-tools/pypi-publish/commit/a020065d46cd15a4db28c06abe013448b7b6b4d4) - use API token for pypi publish instead of user/pass *(commit by [@NGPixel](https://github.com/NGPixel))*


## [v1.0.7] - 2022-02-03
### Bug Fixes
- [`59b394ae06`](https://github.com/ietf-tools/pypi-publish/commit/59b394ae06696c0f0e6ca4b508692cebb931a058) - clean temp dir before and after publish


## [v1.0.6] - 2022-02-03
### New Features
- [`efe26f8723`](https://github.com/ietf-tools/pypi-publish/commit/efe26f8723323d24363a87c6b3d24693d8bfb12b) - check version on pypi before publish


## [v1.0.5] - 2022-02-02
### New Features
- [`e13fb19e8b`](https://github.com/ietf-tools/pypi-publish/commit/e13fb19e8b0fa52c8c651c269cb06b28c3509f93) - add command arguments support


## [v1.0.4] - 2022-02-02
### New Features
- [`9f4449896c`](https://github.com/ietf-tools/pypi-publish/commit/9f4449896c22630580f751d80025ec52b4b61a87) - add gpg identity prompt


## [v1.0.3] - 2022-02-02
### Bug Fixes
- [`6c9ac927e0`](https://github.com/ietf-tools/pypi-publish/commit/6c9ac927e0cbd8b978bec3ba6f4f5dc7a4e0ff29) - ask for target pypi before credentials


## [v1.0.2] - 2022-02-01
### Bug Fixes
- [`e583acb93a`](https://github.com/ietf-tools/pypi-publish/commit/e583acb93a3d919b21621707df6eb9cf2eee76b1) - gh auth additional info + extra input validation


## [v1.0.1] - 2022-02-01
### Bug Fixes
- [`924020d6a3`](https://github.com/ietf-tools/pypi-publish/commit/924020d6a33c194206b9b769a4279e23efca719e) - enforce node 16 or later engine

[v1.0.1]: https://github.com/ietf-tools/pypi-publish/compare/v1.0.0...v1.0.1
[v1.0.2]: https://github.com/ietf-tools/pypi-publish/compare/v1.0.1...v1.0.2
[v1.0.3]: https://github.com/ietf-tools/pypi-publish/compare/v1.0.2...v1.0.3
[v1.0.4]: https://github.com/ietf-tools/pypi-publish/compare/v1.0.3...v1.0.4
[v1.0.5]: https://github.com/ietf-tools/pypi-publish/compare/v1.0.4...v1.0.5
[v1.0.6]: https://github.com/ietf-tools/pypi-publish/compare/v1.0.5...v1.0.6
[v1.0.7]: https://github.com/ietf-tools/pypi-publish/compare/v1.0.6...v1.0.7

[v1.1.0]: https://github.com/ietf-tools/pypi-publish/compare/v1.0.7...v1.1.0

index.js
const { Octokit } = require('octokit')
const { createOAuthDeviceAuth } = require('@octokit/auth-oauth-device')
const inquirer = require('inquirer')
const open = require('open')
const fs = require('fs-extra')
const os = require('os')
const path = require('path')
const pipeline = require('stream/promises').pipeline
const spawn = require('child_process').spawn
const org = 'ietf-tools'

inquirer.registerPrompt('search-list', require('inquirer-search-list'))

async function main () {
  console.info('===========================')
  console.info('IETF Python Publishing Tool')
  console.info('===========================\n')

  const ora = (await import('ora')).default
  const clipboardy = (await import('clipboardy')).default
  const got = (await import('got')).default

  const optsPrompt = await inquirer.prompt([
    {
      type: 'input',
      name: 'python',
      message: 'Enter path to Python executable:',
      default: process.env.PYTHONHOME || process.env.PYTHON || '',
      validate (v) {
        return (v && v.length > 0) || 'Enter a valid Python path'
      }
    },
    {
      type: 'input',
      name: 'user',
      message: 'Enter your PyPI username:',
      validate (v) {
        return (v && v.length > 0) || 'Enter a valid PyPI username'
      }
    },
    {
      type: 'password',
      name: 'pass',
      message: 'Enter your PyPI password:',
      validate (v) {
        return (v && v.length > 0) || 'Enter a PyPI password'
      }
    },
    {
      type: 'list',
      name: 'pypi',
      message: 'Select the target PyPI repository:',
      default: 0,
      choices: ['pypi', 'testpypi']
    }
  ])
  if (!optsPrompt?.python) {
    console.error('No Python path entered. Exiting...')
    process.exit(1)
  }

  const spinnerAuth = ora('Waiting for GitHub authentication to complete...')

  const gh = new Octokit({
    userAgent: 'ietf-pypi-publish',
    authStrategy: createOAuthDeviceAuth,
    auth: {
      clientId: 'e9642b43d2c36ba005b8',
      clientType: 'oauth-app',
      scopes: ['public_repo'],
      onVerification(verif) {
        console.info('\nOpen in your browser: %s', verif.verification_uri)
        console.info('Enter code: %s\n', verif.user_code)
        spinnerAuth.start()
        try {
          clipboardy.writeSync(verif.user_code)
          open(verif.verification_uri)
        } catch (err) {}
      }
    }
  })

  await gh.auth({ type: 'oauth' })
  spinnerAuth.succeed('Authenticated.')

  // -> Fetch GitHub Repos

  const spinnerFetchRepos = ora('Fetching list of GitHub repositories...').start()
  let repos = []
  try {
    const reposRaw = await gh.rest.repos.listForOrg({
      org: org,
      type: 'public',
      sort: 'updated',
      direction: 'desc',
      per_page: 100
    })
    repos = reposRaw?.data?.filter(r => !r.archived && !r.disabled && r.name !== '.github').map(r => r.name).sort() ?? []
  } catch (err) {
    spinnerFetchRepos.fail('Failed to fetch list of GitHub repositories!')
    console.error(err.message)
    process.exit(1)
  }
  spinnerFetchRepos.succeed(`Fetched ${repos.length} most recently updated GitHub repositories.`)

  // -> Select GitHub Repo to use

  const repoPrompt = await inquirer.prompt([
    {
      type: 'search-list',
      name: 'repo',
      message: 'Select the GitHub repository to use:',
      choices: repos
    }
  ])
  if (!repoPrompt?.repo) {
    console.error('Invalid or no repository selected. Exiting...')
    process.exit(1)
  }

  // -> Fetch GitHub releases

  const spinnerFetchReleases = ora('Fetching list of GitHub repositories...').start()
  let releases = []
  try {
    const releasesRaw = await gh.graphql(`
      query lastReleases ($owner: String!, $repo: String!) {
        repository (owner: $owner, name: $repo) {
          releases(first: 10, orderBy: { field: CREATED_AT, direction: DESC }) {
            nodes {
              author {
                login
              }
              createdAt
              id
              name
              releaseAssets (first: 100) {
                nodes {
                  downloadUrl
                  name
                  size
                  id
                  url
                }
              }
              tag {
                name
              }
              url
              isDraft
              isLatest
              isPrerelease
            }
          }
        }
      }
    `, {
      owner: org,
      repo: repoPrompt.repo
    })
    releases = releasesRaw?.repository?.releases?.nodes ?? []
  } catch (err) {
    spinnerFetchReleases.fail('Failed to fetch list of releases!')
    console.error(err.message)
    process.exit(1)
  }
  spinnerFetchReleases.succeed(`Fetched ${releases.length} most recent releases.`)

  // -> Select release to use

  const releasePrompt = await inquirer.prompt([
    {
      type: 'list',
      name: 'release',
      message: 'Select the release to publish:',
      choices: releases.map(r => r.name),
      default: 0
    }
  ])
  if (!releasePrompt?.release) {
    console.error('Invalid or no release selected. Exiting...')
    process.exit(1)
  }
  const release = releases.filter(r => r.name === releasePrompt.release)[0]

  // -> Check for python dist packages

  if (release.releaseAssets.nodes.map(a => a.name).filter(a => a.endsWith('.tar.gz')).length < 1) {
    console.error('Could not find any Python distribution type asset. Make sure the release has a build attached. Exiting...')
    process.exit(1)
  }

  // -> Create temp dir

  const spinnerCreateDir = ora('Downloading release assets...').start()
  let distdir = null
  try {
    tempdir = path.join(os.tmpdir(), 'ietf-pypi-publish')
    distdir = path.join(tempdir, 'dist')
    await fs.ensureDir(distdir)
    spinnerCreateDir.succeed(`Created temp directory: ${distdir}`)
  } catch (err) {
    spinnerCreateDir.fail('Failed to create temp directory.')
    console.error(err.message)
    process.exit(1)
  }

  // -> Download release assets

  const spinnerDownloadAssets = ora({ text: 'Downloading release assets...', spinner: 'arrow3' }).start()
  let assetDownloaded = 0
  for (const asset of release.releaseAssets.nodes) {
    spinnerDownloadAssets.text = `Downloading asset ${asset.name}...`
    try {
      await pipeline(
        got.stream(asset.url),
        fs.createWriteStream(path.join(distdir, asset.name))
      )
      assetDownloaded++
    } catch (err) {
      spinnerCreateDir.fail(`Failed to download asset ${asset.name}.`)
      console.error(err.message)
      process.exit(1)
    }
  }
  spinnerDownloadAssets.succeed(`Downloaded ${assetDownloaded} assets.`)

  // -> Install Twine

  const spinnerInstallTwine = ora('Installing Twine...').start()
  const errorsInstall = []
  try {
    const proc = spawn(optsPrompt?.python, ['-m', 'pip', 'install', 'twine'], {
      cwd: tempdir,
      windowsHide: true,
      timeout: 1000 * 60 * 5
    })
    proc.stderr.on('data', data => {
      errorsInstall.push(data.toString('utf8'))
    })
    await new Promise((resolve, reject) => {
      proc.on('exit', code => {
        if (code > 0) {
          reject(new Error(errorsInstall.join(', ')))
        } else {
          resolve()
        }
      })
    })
  } catch (err) {
    spinnerInstallTwine.fail('Failed to install Twine.')
    console.error(err.message)
    process.exit(1)
  }
  spinnerInstallTwine.succeed('Installed Twine successfully.')

  // -> Run Twine

  const spinnerRunTwine = ora('Publishing package using Twine...').start()
  const errorsRun = []
  try {
    const proc = spawn(optsPrompt.python, ['-m', 'twine', 'upload', '--verbose', '--sign', 'dist/*'], {
      cwd: tempdir,
      windowsHide: true,
      timeout: 1000 * 60 * 5,
      env: {
        ...process.env,
        TWINE_USERNAME: optsPrompt.user,
        TWINE_PASSWORD: optsPrompt.pass,
        TWINE_REPOSITORY_URL: optsPrompt.pypi === 'pypi' ? 'https://upload.pypi.org/legacy/' : 'https://test.pypi.org/legacy/'
      }
    })
    proc.stderr.on('data', data => {
      errorsRun.push(data.toString('utf8'))
    })
    await new Promise((resolve, reject) => {
      proc.on('exit', code => {
        if (code > 0) {
          reject(new Error(errorsRun.join(', ')))
        } else {
          resolve()
        }
      })
    })
  } catch (err) {
    spinnerRunTwine.fail('Failed to publish package.')
    console.error(err.message)
    process.exit(1)
  }
  spinnerRunTwine.succeed('Published package successfully.')

  process.exit(0)
}

main()

cli.js
#!/usr/bin/env node

const { Octokit } = require('octokit')
const { createOAuthDeviceAuth } = require('@octokit/auth-oauth-device')
const inquirer = require('inquirer')
const open = require('open')
const fs = require('fs-extra')
const os = require('os')
const path = require('path')
const pipeline = require('stream/promises').pipeline
const spawn = require('child_process').spawn
const org = 'ietf-tools'

inquirer.registerPrompt('search-list', require('inquirer-search-list'))

async function main () {
  const argv = require('yargs')
    .scriptName('pypi-publish')
    .usage('$0 [args]')
    .options({
      't': {
        alias: 'target',
        describe: 'Target PyPI repository',
        choices: ['pypi', 'testpypi'],
        type: 'string'
      },
      'a': {
        alias: 'token',
        describe: 'PyPI API Token',
        type: 'string'
      },
      'i': {
        alias: 'identity',
        describe: 'GPG identity to use for package signing',
        type: 'string'
      },
      'p': {
        alias: 'project',
        describe: 'GitHub project (repository) to publish from',
        type: 'string'
      },
      'r': {
        alias: 'release',
        describe: 'GitHub release to publish',
        type: 'string'
      },
      'python-path': {
        describe: 'Path to Python executable',
        type: 'string'
      }
    })
    .help()
    .alias('h', 'help')
    .alias('v', 'version')
    .epilogue('All arguments are optional and will be prompted if not provided.')
    .argv

  console.info('===========================')
  console.info('IETF Python Publishing Tool')
  console.info('===========================\n')

  const ora = (await import('ora')).default
  const clipboardy = (await import('clipboardy')).default
  const got = (await import('got')).default
  const chalk = (await import('chalk')).default

  const optsPrompt = await inquirer.prompt([
    {
      type: 'input',
      name: 'python',
      message: 'Enter path to Python executable:',
      default: process.env.PYTHONHOME || process.env.PYTHON || '',
      validate (v) {
        return (v && v.length > 1) || 'Enter a valid Python path'
      }
    },
    {
      type: 'list',
      name: 'pypi',
      message: 'Select the target PyPI repository:',
      default: 0,
      choices: ['pypi', 'testpypi']
    },
    {
      type: 'password',
      name: 'token',
      message: 'Enter your PyPI API token:',
      validate (v) {
        return (v && v.length > 3 && v.startsWith('pypi-')) || 'Enter a PyPI API token'
      }
    },
    {
      type: 'input',
      name: 'gpgidentity',
      message: 'Enter the GPG identity to use for signing (leave empty for default):'
    }
  ], {
    ...argv.pythonPath && { python: argv.pythonPath },
    ...argv.t && { pypi: argv.t },
    ...argv.a && { token: argv.a },
    ...argv.i && { gpgidentity: argv.i }
  })
  if (!optsPrompt?.python) {
    console.error(chalk.redBright('No Python path entered. Exiting...'))
    process.exit(1)
  }

  const spinnerAuth = ora('Waiting for GitHub authentication to complete...')

  const gh = new Octokit({
    userAgent: 'ietf-pypi-publish',
    authStrategy: createOAuthDeviceAuth,
    auth: {
      clientId: 'e9642b43d2c36ba005b8',
      clientType: 'oauth-app',
      scopes: ['public_repo'],
      onVerification(verif) {
        console.info(`
Open in your browser: ${chalk.underline.green(verif.verification_uri)}
Enter code: ${chalk.bold.yellowBright(verif.user_code)}

${chalk.italic.grey('(The code has already been copied to your clipboard for convenience.)')}
        `)
        spinnerAuth.start()
        try {
          clipboardy.writeSync(verif.user_code)
          open(verif.verification_uri)
        } catch (err) {}
      }
    }
  })

  await gh.auth({ type: 'oauth' })
  spinnerAuth.succeed('Authenticated to GitHub.')

  // -> Fetch GitHub Repos

  const spinnerFetchRepos = ora('Fetching list of GitHub repositories...').start()
  let repos = []
  try {
    const reposRaw = await gh.rest.repos.listForOrg({
      org: org,
      type: 'public',
      sort: 'updated',
      direction: 'desc',
      per_page: 100
    })
    repos = reposRaw?.data?.filter(r => !r.archived && !r.disabled && r.name !== '.github').map(r => r.name).sort() ?? []
  } catch (err) {
    spinnerFetchRepos.fail('Failed to fetch list of GitHub repositories!')
    console.error(chalk.redBright(err.message))
    process.exit(1)
  }
  spinnerFetchRepos.succeed(`Fetched ${repos.length} most recently updated GitHub repositories.`)

  // -> Select GitHub Repo to use

  let repo = null
  if (argv.p) {
    if (repos.includes(argv.p)) {
      repo = argv.p
      ora(`Using GitHub repository: ${repo}`).succeed()
    } else {
      console.warn(chalk.redBright('Invalid GitHub repository provided.'))
    }
  }

  if (!repo) {
    let repoPrompt = await inquirer.prompt([
      {
        type: 'search-list',
        name: 'repo',
        message: 'Select the GitHub repository to use:',
        choices: repos
      }
    ])
    if (!repoPrompt?.repo) {
      console.error(chalk.redBright('Invalid or no repository selected. Exiting...'))
      process.exit(1)
    }
    repo = repoPrompt.repo
  }

  // -> Fetch GitHub releases

  const spinnerFetchReleases = ora('Fetching list of GitHub repositories...').start()
  let releases = []
  try {
    const releasesRaw = await gh.graphql(`
      query lastReleases ($owner: String!, $repo: String!) {
        repository (owner: $owner, name: $repo) {
          releases(first: 10, orderBy: { field: CREATED_AT, direction: DESC }) {
            nodes {
              author {
                login
              }
              createdAt
              id
              name
              releaseAssets (first: 100) {
                nodes {
                  downloadUrl
                  name
                  size
                  id
                  url
                }
              }
              tag {
                name
              }
              url
              isDraft
              isLatest
              isPrerelease
            }
          }
        }
      }
    `, {
      owner: org,
      repo: repo
    })
    releases = releasesRaw?.repository?.releases?.nodes ?? []
  } catch (err) {
    spinnerFetchReleases.fail('Failed to fetch list of releases!')
    console.error(chalk.redBright(err.message))
    process.exit(1)
  }
  if (releases.length > 0) {
    spinnerFetchReleases.succeed(`Fetched ${releases.length} most recent releases.`)
  } else {
    spinnerFetchReleases.fail('This project has no release! Exiting...')
    process.exit(1)
  }

  // -> Select release to use

  let releaseName = null
  if (argv.r) {
    if (releases.map(r => r.name).includes(argv.r)) {
      releaseName = argv.r
      ora(`Using GitHub release: ${releaseName}`).succeed()
    } else {
      console.warn(chalk.redBright('Invalid GitHub release provided.'))
    }
  }

  if (!releaseName) {
    const releasePrompt = await inquirer.prompt([
      {
        type: 'list',
        name: 'release',
        message: 'Select the release to publish:',
        choices: releases.map(r => r.name),
        default: 0
      }
    ])
    if (!releasePrompt?.release) {
      console.error(chalk.redBright('Invalid or no release selected. Exiting...'))
      process.exit(1)
    }
    releaseName = releasePrompt.release
  }
  const release = releases.filter(r => r.name === releaseName)[0]

  // -> Check for python dist packages

  if (release.releaseAssets.nodes.map(a => a.name).filter(a => a.endsWith('.tar.gz')).length < 1) {
    console.error(chalk.redBright('Could not find any Python distribution type asset. Make sure the release has a build attached. Exiting...'))
    process.exit(1)
  }

  // -> Check for existing version on PyPI

  const spinnerCheckExistingVer = ora('Checking for existing version on PyPI...').start()
  const pypiHost = optsPrompt.pypi === 'pypi' ? 'pypi.org' : 'test.pypi.org'
  try {
    await got({
      url: `https://${pypiHost}/pypi/${repo}/${release.name}/json`
    }).json()
    spinnerCheckExistingVer.fail(`Version ${release.name} already exists on ${pypiHost}. Cannot overwrite an existing version! Exiting...`)
    process.exit(1)
  } catch (err) {
    spinnerCheckExistingVer.succeed(`Version ${release.name} does not exist yet on ${pypiHost}.`)
  }

  // -> Create temp dir

  const spinnerCreateDir = ora('Downloading release assets...').start()
  let tempdir = null
  let distdir = null
  try {
    tempdir = path.join(os.tmpdir(), 'ietf-pypi-publish')
    distdir = path.join(tempdir, 'dist')
    await fs.emptyDir(tempdir)
    await fs.ensureDir(distdir)
    spinnerCreateDir.succeed(`Created temp directory: ${distdir}`)
  } catch (err) {
    spinnerCreateDir.fail('Failed to create temp directory.')
    console.error(chalk.redBright(err.message))
    process.exit(1)
  }

  // -> Download release assets

  const spinnerDownloadAssets = ora({ text: 'Downloading release assets...', spinner: 'arrow3' }).start()
  let assetDownloaded = 0
  for (const asset of release.releaseAssets.nodes) {
    spinnerDownloadAssets.text = `Downloading asset ${asset.name}...`
    try {
      await pipeline(
        got.stream(asset.url),
        fs.createWriteStream(path.join(distdir, asset.name))
      )
      assetDownloaded++
    } catch (err) {
      spinnerCreateDir.fail(`Failed to download asset ${asset.name}.`)
      console.error(chalk.redBright(err.message))
      process.exit(1)
    }
  }
  spinnerDownloadAssets.succeed(`Downloaded ${assetDownloaded} assets.`)

  // -> Install Twine

  const spinnerInstallTwine = ora('Installing Twine...').start()
  const errorsInstall = []
  try {
    const proc = spawn(optsPrompt?.python, ['-m', 'pip', 'install', 'twine'], {
      cwd: tempdir,
      windowsHide: true,
      timeout: 1000 * 60 * 5
    })
    proc.stderr.on('data', data => {
      errorsInstall.push(data.toString('utf8'))
    })
    await new Promise((resolve, reject) => {
      proc.on('exit', code => {
        if (code > 0) {
          reject(new Error(errorsInstall.join(', ')))
        } else {
          resolve()
        }
      })
    })
  } catch (err) {
    spinnerInstallTwine.fail('Failed to install Twine.')
    console.error(chalk.redBright(err.message))
    process.exit(1)
  }
  spinnerInstallTwine.succeed('Installed Twine successfully.')

  // -> Last prompt check before publishing...

  const confirmPrompt = await inquirer.prompt([
    {
      type: 'confirm',
      name: 'go',
      message: `Proceed with publishing package ${repo}: ${release.name} to ${optsPrompt.pypi}?`,
      default: false
    }
  ])
  if (!confirmPrompt?.go) {
    console.error(chalk.redBright('Publishing aborted by the user. Exiting...'))
    process.exit(1)
  }

  // -> Run Twine

  const spinnerRunTwine = ora('Publishing package using Twine...').start()
  const errorsRun = []
  try {
    const twineParams = ['-m', 'twine', 'upload', '--verbose', '--sign']
    if (optsPrompt.gpgidentity) {
      twineParams.push('--identity')
      twineParams.push(optsPrompt.gpgidentity)
    }
    twineParams.push('dist/*')
    const proc = spawn(optsPrompt.python, twineParams, {
      cwd: tempdir,
      windowsHide: true,
      timeout: 1000 * 60 * 5,
      env: {
        ...process.env,
        TWINE_USERNAME: '__token__',
        TWINE_PASSWORD: optsPrompt.token,
        TWINE_REPOSITORY_URL: optsPrompt.pypi === 'pypi' ? 'https://upload.pypi.org/legacy/' : 'https://test.pypi.org/legacy/'
      }
    })
    proc.stderr.on('data', data => {
      errorsRun.push(data.toString('utf8'))
    })
    await new Promise((resolve, reject) => {
      proc.on('exit', code => {
        if (code > 0) {
          reject(new Error(errorsRun.join(', ')))
        } else {
          resolve()
        }
      })
    })
  } catch (err) {
    spinnerRunTwine.fail('Failed to publish package.')
    console.error(chalk.redBright(err.message))
    process.exit(1)
  }
  spinnerRunTwine.succeed('Published package successfully.')

  // -> Clean up temp directory

  try {
    await fs.emptyDir(tempdir)
  } catch (err) {
    console.error(chalk.yellow(`Unable to clean temp folder ${tempdir}`))
  }

  process.exit(0)
}

main()
